### PR TITLE
Lead 86/system status alerts

### PR DESCRIPTION
**Pull request by hutnikau**
Added the script which check if there is any system status degradation since last check.

To test it do the following steps:
1. Make sure that mathJax installed. 
1.a) "Install" if it's required with the following command: `mkdir -p taoQtiItem/views/js/mathjax/ && touch taoQtiItem/views/js/mathjax/MathJax.js`
2. Run degradations checker: `php index.php '\oat\taoSystemStatus\scripts\tools\CheckDegradations' -p default_kv` (no degradations should be found)
3. Delete or rename mathjax file (`taoQtiItem/views/js/mathjax/MathJax.js`)
4. Run degradations checker: `php index.php '\oat\taoSystemStatus\scripts\tools\CheckDegradations' -p default_kv` (Service degradations found ...)

### CODE REVIEW

File: model/Report/ReportComparator.php [hunk 1]
```diff
@@ -0,0 +1,139 @@
+<?php
```
Comment (gitromba) [inline model/Report/ReportComparator.php:1]: Would be better with strict types.

```suggestion
<?php

declare(strict_types=1);
```

File: model/Report/ReportComparator.php [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        foreach ($report->getChildren() as $childReport) {
+            if ($childReport->getType() === Report::TYPE_INFO) {
+                continue;
+            }
+            $result[$childReport->getData()[CheckInterface::PARAM_CHECK_ID]] = $childReport;
+        }
+        ksort($result);
+        return $result;
+    }
+}
```
Comment (gitromba) [inline model/Report/ReportComparator.php]: Missing new line.

File: scripts/tools/CheckDegradations.php [hunk 3]
```diff
@@ -0,0 +1,115 @@
+<?php
+/**
```
Comment (gitromba) [inline scripts/tools/CheckDegradations.php:4]: You might want to add strict type declaration for every new PHP file.

File: model/Report/ReportComparator.php [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+        if ($report->hasChildren()) {
+            $report->setType(Report::TYPE_SUCCESS);
+            $report->setMessage('Service restorations found');
+        }
+
+        return $report;
+    }
+
+    private function compare($trendComparator)
```
Comment (gitromba) [inline model/Report/ReportComparator.php]: You may want to benefit from being type safe.

```suggestion
    private function compare(callable $trendComparator): Result
```

File: model/Report/ReportComparator.php [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+        return $result;
+    }
+
+    /**
+     * @param Report $oldReport
+     * @param Report $newReports
+     * @return int
+     */
+    private function getTrend(Report $oldReport, Report $newReports)
```
Comment (gitromba) [inline model/Report/ReportComparator.php]: Explicit return type is recommended:
```suggestion
    private function getTrend(Report $oldReport, Report $newReports): int
```

File: scripts/tools/CheckDegradations.php [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+     * @return array[]
+     */
+    public function provideOptions()
+    {
+        return [
+            'persistence' => [
+                'prefix' => 'p',
+                'longPrefix' => 'persistence',
+                'required' => true,
+                'description' => 'The KeyValue persistence where you want store results of previous check',
```
Comment (gitromba) [inline scripts/tools/CheckDegradations.php]: As a newbie with this code, it would greatly help my understanding if I could see the available values, or an example.

File: scripts/tools/CheckDegradations.php [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        $result = $comparator->getDegradations();
+
+        $this->getPersistence()->set(self::KV_PERSISTENCE_KEY, json_encode($report));
+        return $result;
+    }
+
+    /**
+     * @return Report|null
+     */
+    private function getPreviousReport()
```
Comment (gitromba) [inline scripts/tools/CheckDegradations.php]: Explicit return type is better than docblock.

**Comment by gitromba** [comment]
- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [ ] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful
- [ ] Changelog is updated according to changes (if applicable)
- [ ] Documentation is updated according to changes (if applicable)

File: model/Report/ReportComparator.php [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+     * @var Report
+     */
+    private $oldReport;
+
+    /**
+     * @var Report
+     */
+    private $newReport;
+
+    const REPORTS_MAP = [
```
Comment (gitromba) [inline model/Report/ReportComparator.php]: `INFO` type to be handled?

File: model/Report/ReportComparator.php [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    /**
+     * @param Report $oldReport
+     * @param Report $newReports
+     * @return int
+     */
+    private function getTrend(Report $oldReport, Report $newReports)
+    {
+        return self::REPORTS_MAP[$oldReport->getType()] - self::REPORTS_MAP[$newReports->getType()];
```
Comment (gitromba) [inline model/Report/ReportComparator.php]: This will run on error for `INFO` reports.

File: model/Report/ReportComparator.php [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+     * @return Report[]
+     */
+    private function mapReportsById(Report $report): array
+    {
+        $result = [];
+        foreach ($report->getChildren() as $childReport) {
+            if ($childReport->getType() === Report::TYPE_INFO) {
+                continue;
+            }
+            $result[$childReport->getData()[CheckInterface::PARAM_CHECK_ID]] = $childReport;
```
Comment (gitromba) [inline model/Report/ReportComparator.php]: `Report::getData()` can be anything, should be checked whether it's an array before being treated as one.

File: model/Report/ReportComparator.php [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+
+    /**
+     * @param Report $report
+     * @return Report[]
+     */
+    private function mapReportsById(Report $report): array
+    {
+        $result = [];
+        foreach ($report->getChildren() as $childReport) {
+            if ($childReport->getType() === Report::TYPE_INFO) {
```
Comment (gitromba) [inline model/Report/ReportComparator.php]: I wouldn't expect the `INFO` reports to be filtered out in a method named `mapReportsById`. `INFO` reports not being relevant seems to be an important detail, it should be more visible.

File: scripts/tools/CheckDegradations.php [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+    {
+        return 'Script checks if there are any degradations in the system status since last launch';
+    }
+
+    public function run()
+    {
+        $report = $this->getSystemStatusService()->check();
+        $previousReport = $this->getPreviousReport();
+        if (!$previousReport) {
+            return new Report('No previous report found');
```
Comment (gitromba) [inline scripts/tools/CheckDegradations.php]: The first parameter must be a `type`. Or you could use the `Report::createInfo()` named constructor.

File: model/Report/ReportComparator.php (hunk 8 above)
Comment (gitromba) [inline model/Report/ReportComparator.php]: After checking the CR in more depth, I've realized this map is about assigning weight to each report type. Hence I recommend a more expressive name, like `REPORT_WEIGHT_MAP`

File: model/Report/ReportComparator.php [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+
+        return $report;
+    }
+
+    private function compare($trendComparator)
+    {
+        $oldReports = $this->mapReportsById($this->oldReport);
+        $newReports = $this->mapReportsById($this->newReport);
+
+        $result = new Report(Report::TYPE_SUCCESS, 'No changes found');
```
Comment (gitromba) [inline model/Report/ReportComparator.php]: The `No changes found` message seems to be irrelevant and might get inconsistent after the coming lines with the object's state. I advise removing it:
```suggestion
        $result = Report::createInfo();
```

File: model/Report/ReportComparator.php [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    public function getRestorations(): Report
+    {
+        $report = $this->compare(function ($trend) {
+            return $trend > 0;
+        });
+
+        if ($report->hasChildren()) {
+            $report->setType(Report::TYPE_SUCCESS);
```
Comment (gitromba) [inline model/Report/ReportComparator.php:89]: I would have gone with `info`. It makes me puzzled when restoration report is always success, degradation is always warning. It just gives _information_ on the status.

File: test/model/Report/ReportComparatorTest.php [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+        $comparator = new ReportComparator($oldReport, $oldReport);
+        $this->assertFalse($comparator->getDegradations()->hasChildren());
+    }
+
+    public function testGetRestorations()
+    {
+        $oldReport = $this->getInitialReport();
+        $newReport = $this->getNewReport();
+
+        $expected = new Report(Report::TYPE_SUCCESS, 'Service restorations found');
```
Comment (gitromba) [inline test/model/Report/ReportComparatorTest.php:61]: Can I expect a `warning` report if no restoration is found? `info` would be more obvious IMO.

File: test/model/Report/ReportComparatorTest.php [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    private function getReport($type, $message, $id = null): Report
+    {
+        return new Report($type, $message, [
+            CheckInterface::PARAM_CHECK_ID => $id,
+            CheckInterface::PARAM_DETAILS => $id,
+        ]);
+    }
+}
```
Comment (gitromba) [inline test/model/Report/ReportComparatorTest.php:104]: Missing new line.

**Comment by gitromba** [comment]
Good job @hutnikau!

I would extend the unit tests to make sure `info` reports are well-handled.

Locally testing the solution has brought to surface some issues:
- https://github.com/oat-sa/extension-tao-system-status/pull/76#discussion_r500535977 is a bug and must be fixed.
- unless the persistence layer is somehow filled up manually with the initial state, the `CheckDegradations` script action will never be able succeed due to this conditional: https://github.com/oat-sa/extension-tao-system-status/pull/76/files#diff-5817b8ef1a3f54f3d64b2dcee59de34cR70. How do you plan to overcome this?

**Review by gitromba — CHANGES_REQUESTED** [review]
Good job in overall! There are some issues to be addressed, you find them in my previous comments.

File: model/Report/ReportComparator.php (hunk 14 above)
Comment (hutnikau) [reply model/Report/ReportComparator.php:89]: In that context no warnings or errors means success. I would like to leave it as is as the type of report may be used as a marker that something happened (good or bad).

File: model/Report/ReportComparator.php (hunk 13 above)
Comment (hutnikau) [reply model/Report/ReportComparator.php]: fixed

File: model/Report/ReportComparator.php (hunk 9 above)
Comment (hutnikau) [reply model/Report/ReportComparator.php]: fixed

File: scripts/tools/CheckDegradations.php (hunk 3 above)
Comment (hutnikau) [reply scripts/tools/CheckDegradations.php:4]: fixed

**Comment by gitromba** [comment]
Thank you for the changes, Aleh. 

Unit tests are green 👌 
The described functional test scenario however didn't pass :/

```
❯ docker exec -it tao-phpfpm /bin/sh
# php index.php '\oat\taoSystemStatus\scripts\tools\RunInstanceCheck'
sh: 1: ps: not found
All Systems Operational:
  MathJax is configured correctly
  All messages.json exist
  TAO has permissions to write into 'config' and 'data' folders
# php index.php '\oat\taoSystemStatus\scripts\tools\CheckDegradations' -p default_kv
No previous report found
# rm -f taoQtiItem/views/js/mathjax/MathJax.js
# php index.php '\oat\taoSystemStatus\scripts\tools\RunInstanceCheck'
sh: 1: ps: not found
Partially Degraded Service:
  MathJax.js does not exist
  All messages.json exist
  TAO has permissions to write into 'config' and 'data' folders
# php index.php '\oat\taoSystemStatus\scripts\tools\CheckDegradations' -p default_kv
No previous report found
```

The degradation was not found, despite Mathjax was removed after the first instance check.

**Review by gitromba — CHANGES_REQUESTED** [review]
Please check my last comment, degradations are not found.

**Comment by hutnikau** [comment]
@gitromba i changed degradations checker in order to run instance check automatically. So now there is no need to run RunInstanceCheck. Please see updated PR description.

**Comment by gitromba** [comment]
I've applied your latest changes @hutnikau in my local environment. Degradation by deleting the mathjax file is **not** found.

I did a clean install with docker stack:
1. clone package-tao repo
2. checkout develop
3. extend composer with:
-- "oat-sa/extension-tao-system-status": "dev-LEAD-86/system_status_alerts",
-- "oat-sa/extension-tao-scheduler": "dev-develop"
-- "oat-sa/extension-tao-task-queue": "dev-develop"
4. composer install
5. add to `systemStatus` extension to `taoTest` manifest ('taoSystemStatus' => '>=0.13.0')
6. taoDockerize install

Result:
```
# php index.php '\oat\taoSystemStatus\scripts\tools\CheckDegradations' -p default_kv
sh: 1: ps: not found
No previous report found
# rm -f taoQtiItem/views/js/mathjax/MathJax.js
# php index.php '\oat\taoSystemStatus\scripts\tools\CheckDegradations' -p default_kv
sh: 1: ps: not found
No previous report found
# php index.php '\oat\taoSystemStatus\scripts\tools\RunInstanceCheck'
sh: 1: ps: not found
Partially Degraded Service:
  MathJax.js does not exist
  All messages.json exist
  TAO has permissions to write into 'config' and 'data' folders
# php index.php '\oat\taoSystemStatus\scripts\tools\CheckDegradations' -p default_kv
sh: 1: ps: not found
No previous report found
```

~~I hope it's just my local setup, but I recommend having it checked by others well.~~

EDIT:
I've checked the code, and my result is not false negative unfortunately. `CheckDegradations::getPreviousReport()` always return null as there's **no** degradation report available in the persistence layer. Please re-visit this part and apply a fix.

**Comment by hutnikau** [comment]
@gitromba I pushed the fix. Please test it again.

**Review by gitromba — APPROVED** [review]
Thank you for the fix @hutnikau. Looks good 👊 .

```
❯ docker exec -it tao-phpfpm /bin/sh
# mkdir -p taoQtiItem/views/js/mathjax/ && touch taoQtiItem/views/js/mathjax/MathJax.js
#
# mkdir -p taoQtiItem/views/js/mathjax/ && touch taoQtiItem/views/js/mathjax/MathJax.js
# php index.php '\oat\taoSystemStatus\scripts\tools\CheckDegradations' -p default_kv
sh: 1: ps: not found
No previous report found
# rm -f taoQtiItem/views/js/mathjax/MathJax.js
# php index.php '\oat\taoSystemStatus\scripts\tools\CheckDegradations' -p default_kv
sh: 1: ps: not found
Service degradations found
  MathJax.js does not exist
# php index.php '\oat\taoSystemStatus\scripts\tools\CheckDegradations' -p default_kv
sh: 1: ps: not found
No changes found
# php index.php '\oat\taoSystemStatus\scripts\tools\CheckDegradations' -p default_kv
sh: 1: ps: not found
No changes found
```

**Review by bugalot — APPROVED** [review]
Sounds great! Thanks!
